### ISSUE TITLE
AR 4.1.10.rc1 introduced bug on uniqueness validation with scope

### ISSUE THREAD

**Issue by bquorning** [rails/rails#19051]
When a model has uniqueness validation scoped to an attribute whose value is nil (or an array of attributes of which one is nil), the validation passes for identical values.

See https://gist.github.com/bquorning/460ad070893cbe498df5 for the bug template. Both the tests pass using Rails v4.1.9. The last test fails using Rails v4.1.10.rc1.

I suspect this bug was introduced by #18932 – which wasn’t mentioned in the CHANGELOG :trollface:

**Comment by rafaelfranca** [comment]
cc @sgrif

**Comment by kuldeepaggarwal** [comment]
I think, we need to implement this in similar manner as on master, i.e. user ActiveRecord::Relation#where instead of directly using AREL.

**Comment by sgrif** [comment]
The changes required to do what is happening on master aren't possible on the version of Arel used by 4.1 and it would be non trivial to do so. I'll dig into it today.

**Comment by raouldevil** [comment]
I have replicated this issue: see console output at: https://gist.github.com/raouldevil/301918f2590d7311ed8a

**Comment by kuldeepaggarwal** [comment]
@sgrif: Actually I found the issue and its because we used ArelTable#eq that is creating problem.

```ruby
# when node is nil
table[scope_item].eq(node) # => scope_item = NULL

# you noticed that it is creating wrong sql and that is creating an issue.
```

And I wanted to say that we should use ActiveRecord::Relation#where here.

**Comment by sgrif** [comment]
Fixed in 1e1d0f1

**Comment by bquorning** [comment]
Thanks, @sgrif
